src/address.rs

```diff
@@ -68,66 +77,334 @@ impl Matcher {
+                            }
+                            _ => match_wildcard(remainder, *l, Some(next)),
+                        }
+                    } else {
+                        match_wildcard(remainder, *l, None)
+                    }
+                }
+                AddressPatternComponent::CharacterClass(cc) => match_character_class(remainder, cc),
+                AddressPatternComponent::Choice(s) => match_choice(remainder, s),
+            };
+
+            match result {
+                Ok((i, _)) => remainder = i,
+                Err(_) => return Ok(false), // Component didn't match, goodbye
+            };
+        }
+
+        // Address is only matched if it was consumed entirely
+        if remainder.is_empty() {
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+}
+
+/// Check whether a character is an allowed address character
+/// All printable ASCII characters except for a few special characters are allowed
+fn is_address_character(x: char) -> bool {
+    match x {
+        ' ' | '#' | '*' | ',' | '/' | '?' | '[' | ']' | '{' | '}' => false,
+        c => c > '\x20' && c < '\x7F',
+    }
+}
+
+/// Parser to turn a choice like '{foo,bar}' into a vector containing the choices, like ["foo", "bar"]
+fn pattern_choice(input: &str) -> IResult<&str, Vec<&str>> {
+    delimited(
+        char('{'),
+        separated_list1(tag(","), take_while1(is_address_character)),
+        char('}'),
+    )(input)
+}
+
+/// Parser to recognize a character class like [!a-zA-Z] and return '!a-zA-Z'
+fn pattern_character_class(input: &str) -> IResult<&str, &str> {
+    let inner = pair(
+        // It is important to read the leading negating '!' to make sure the rest of the parsed
+        // character class isn't empty. E.g. '[!]' is not a valid character class.
+        recognize(opt(tag("!"))),
+        take_while1(is_address_character),
+    );
+
+    delimited(char('['), recognize(inner), char(']'))(input)
+}
+
+/// A characters class is defined by a set or range of characters that it matches.
+/// For example, [a-z] matches all lowercase alphabetic letters. It can also contain multiple
+/// ranges, like [a-zA-Z]. Instead of a range you can also directly provide the characters to match,
+/// e.g. [abc123]. You can also combine this with ranges, like [a-z123].
+/// If the first characters is an exclamation point, the match is negated, e.g. [!0-9] will match
+/// anything except numbers.
+#[derive(Debug)]
+struct CharacterClass {
+    pub negated: bool,
+    pub characters: String,
+}
+
+/// Expand a character range like 'a-d' to all the letters contained in the range, e.g. 'abcd'
+/// This is done by converting the characters to their ASCII values and then getting every ASCII
+/// in between.
+fn expand_character_range(first: char, second: char) -> String {
+    let start = first as u8;
+    let end = second as u8;
+
+    let range = if start >= end {
```

Comment:
It would be good if we can detect invalid character ranges when validating the pattern.  We should not try to be clever here and instead fail the parsing.

Reply:
Invalid meaning `start > end`.

Reply:
The specification says:
> two characters separated by a minus sign indicate the range of characters between the given two in ASCII collating sequence.

As I interpreted it, doing something like z-a is just as valid as a-z, and also 0-z and other wild combinations. At least I don't see anything wrong with implementing it like this, although I'd really love to see someone actually use that 😅

Reply:
Alright, but if we support this now we must stick with it.  I just checked what happens for "inversed" ranges and they're just empty: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3512162b21dc42779ac867e8a17889fc

I would suggest to not flip the range and instead use them as-is.  This is a bit about educating users/clients to fix their messages ;)

Reply:
I don't see why we shouldn't support a reversed range, maybe I'm misunderstanding what "collating sequence" means? I thought it just means that it's essentially a range, like `a-d` expanding to `abcd` and `d-a` resulting in `dcba`. I'm fine either way, but I'd just like to be on the same page :)

But if we don't support it we should have it result in a parser error, the same way regex does. That'll make the nom parser a bit more complicated for less functionality.

Reply:
I've extended the parser to verify that ranges are sorted by ASCII-value, so `a-d` is still valid, but `d-a` or `a-a` are not valid anymore. Range flipping is thus also obsolete, and hence removed.